rust/operator-binary/src/controller.rs

```diff
@@ -131,15 +138,84 @@ impl ReconcilerError for Error {
     }
 }
 
+const ENV_S3_ACCESS_KEY: &str = "S3_ACCESS_KEY";
+const ENV_S3_SECRET_KEY: &str = "S3_SECRET_KEY";
+const SECRET_KEY_S3_ACCESS_KEY: &str = "accessKey";
+const SECRET_KEY_S3_SECRET_KEY: &str = "secretKey";
+
 pub async fn reconcile_trino(trino: Arc<TrinoCluster>, ctx: Context<Ctx>) -> Result<Action> {
     tracing::info!("Starting reconcile");
 
     let client = &ctx.get_ref().client;
     let version = trino_version(&trino)?;
 
-    let validated_config =
+    let mut validated_config =
         validated_product_config(&trino, version, &ctx.get_ref().product_config)?;
 
+    let s3_connection_spec = if let Some(s3) = &trino.spec.s3 {
+        Some(
+            s3.resolve(
+                client,
+                Some(
+                    trino
+                        .namespace()
+                        .as_deref()
+                        .context(ObjectHasNoNamespaceSnafu)?,
+                ),
+            )
+            .await
+            .context(ResolveS3ConnectionSnafu)?,
+        )
+    } else {
+        None
+    };
+
+    match &s3_connection_spec {
+        Some(
+            s3_connection_spec @ S3ConnectionSpec {
+                host: Some(_),
+                path_style_access,
+                secret_class,
+                tls,
+                ..
+            },
+        ) => {
+            for role_config in &mut validated_config.values_mut() {
+                for config in role_config.values_mut() {
+                    let hive_properties = config
+                        .entry(PropertyNameKind::File(HIVE_PROPERTIES.to_string()))
+                        .or_default();
+                    hive_properties.insert(
+                        S3_ENDPOINT.to_string(),
+                        s3_connection_spec.endpoint().unwrap(),
+                    );
+                    if secret_class.is_some() {
+                        hive_properties.insert(
+                            S3_ACCESS_KEY.to_string(),
+                            format!("${{ENV:{ENV_S3_ACCESS_KEY}}}"),
+                        );
+                        hive_properties.insert(
+                            S3_SECRET_KEY.to_string(),
+                            format!("${{ENV:{ENV_S3_SECRET_KEY}}}"),
+                        );
+                    }
+                    hive_properties.insert(S3_SSL_ENABLED.to_string(), tls.is_some().to_string());
+                    hive_properties.insert(
+                        S3_PATH_STYLE_ACCESS.to_string(),
+                        path_style_access
+                            .unwrap_or(DEFAULT_PATH_STYLE_ACCESS)
+                            .to_string(),
+                    );
+                }
+            }
+        }
+        Some(S3ConnectionSpec { host: None, .. }) => InvalidS3ConnectionSnafu {
+            reason: "host is missing",
+        }
+        .fail()?,
+        None => (),
+    }
```

Comment:
My preference would be to have that directly in the build config map method, but you can ignore that.
I think it fits there, but you would already stop the reconcile here in case something is wrong and not create services etc. i like that too.

Reply:
My intent was to first gather the complete configuration, to validate it and then to apply it.